axelrod/strategies/dbs.py

```diff
@@ -0,0 +1,354 @@
+        if not self.history:
+            return C
+        
+        if (len(opponent.history)>=2):
+
+            # update history_by_cond
+            # (i.e. update Rp)
+            self.update_history_by_cond(opponent.history)
+     
+            two_moves_ago = (self.history[-2],opponent.history[-2])
+            r_plus = (two_moves_ago,opponent.history[-1])
+            r_minus = (two_moves_ago,({C,D}-{opponent.history[-1]}).pop())
+
+            if r_plus[0] not in self.Rc.keys(): 
+                if self.should_promote(r_plus,self.promotion_threshold):
+                    self.Rc[r_plus[0]]=action_to_int(r_plus[1])
+                    self.violation_counts[r_plus[0]]=0
+                    self.violation_counts[r_plus[0]]=0
+
+            # (if r+ or r- in Rc)
+            if r_plus[0] in self.Rc.keys():
+                to_check = (C if self.Rc[r_plus[0]]==1 else D)
+                # (if r+ in Rc)
+                if r_plus[1]==to_check:
+                    # set the violation count of r+ to 0
+                    self.violation_counts[r_plus[0]]=0
+                # (if r- in Rc)
+                elif r_minus[1]==to_check:
+                    # increment violation count of r-
+                    self.violation_counts[r_plus[0]]+=1
+                    if self.should_demote(r_minus,self.violation_threshold):
+                        self.Rd.update(self.Rc)
+                        self.Rc.clear()
+                        self.violation_counts.clear()
+                        self.v=0
+
+            # r+ in Rc
+            r_plus_in_Rc = (r_plus[0] in self.Rc.keys() and self.Rc[r_plus[0]]==action_to_int(r_plus[1]))
+            # r- in Rd
+            r_minus_in_Rd = (r_minus[0] in self.Rd.keys()  and self.Rd[r_minus[0]]==action_to_int(r_minus[1]))
+
+            if r_minus_in_Rd:
+                self.v+=1
+
+            if self.v > self.reject_threshold or (r_plus_in_Rc and r_minus_in_Rd):
+                self.Rd.clear()
+                self.v = 0
+
+            # compute Rp for conditions that are neither in Rc or Rd
+            Rp = Policy()
+            all_cond = [(C,C),(C,D),(D,C),(D,D)]
+            for outcome in all_cond:
+                if ((outcome not in self.Rc.keys()) and (outcome not in self.Rd.keys())):
+                    # then we need to compute opponent's C answer probability
+                    Rp[outcome] = self.compute_prob_rule(outcome,self.alpha)
+
+            self.Pi=Policy()
+            # algorithm ensure no duplicate keys -> no key overwriting
+            self.Pi.update(self.Rc)
+            self.Pi.update(self.Rd)
+            self.Pi.update(Rp)
+
+        # React to the opponent's last move
+        return MoveGen((self.history[-1],opponent.history[-1]),self.Pi,depth_search_tree=self.tree_depth)
+
+
+# Policy as defined in the article, i.e. a set of (last_move,p) where p is
+# the probability to cooperate in the next move considering last move
+class Policy(dict):
+
+    def __init__(self):
```

Comment:
I have deleted the Policy class. Policies are now represented by a simple dictionary, and there is a create_policy methods to instantiate them.